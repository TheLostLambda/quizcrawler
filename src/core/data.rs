#![allow(clippy::new_ret_no_self)]
use super::logic;
use serde::{Deserialize, Serialize};
use std::{
    cell::RefCell,
    hash::{Hash, Hasher},
    rc::Rc,
    time::SystemTime,
};
use uuid::Uuid;

// The reference-counter (Rc) allows for several structures to own this type at
// once (avoiding messy lifetime tracking). The RefCell allows for the mutable
// borrowing of the underlying question at run-time. This shifts borrow-checking
// to run-time as the static checker can't prove this multiple ownership and
// mutation is safe. No copies means that mutation in one part of the program is
// always seen in the others.
pub type QuestionRef = Rc<RefCell<Question>>;

// I really don't know how I feel about these public fields...
#[derive(Clone, Serialize, Deserialize)]
pub struct Section {
    pub name: String,
    pub questions: Vec<QuestionRef>,
    pub children: Vec<Section>,
}

impl Section {
    // Not sure if I should keep this around...
    pub fn new(name: String, children: Vec<Section>, questions: Vec<Question>) -> Section {
        let questions = questions
            .into_iter()
            .map(|c| Rc::new(RefCell::new(c)))
            .collect();
        Section {
            name,
            children,
            questions,
        }
    }

    pub fn is_parent(&self) -> bool {
        !self.children.is_empty()
    }

    pub fn child_at_path(&self, path: &[impl AsRef<str>]) -> Option<&Section> {
        let mut current = self;
        for name in path {
            // FIXME: What happens when there are sections with the same name?
            current = current.children.iter().find(|c| c.name == name.as_ref())?;
        }
        Some(current)
    }
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum Strictness {
    Exact,
    Trimmed,
    Caseless,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct Question {
    pub id: Uuid,
    pub data: QuestionVariant,
    pub comp_level: Strictness, // FIXME: Should this be moved up to the state machine?
    pub mastery: u8,            // 0-10 (Leitner System)
    pub correct: usize,
    pub seen: usize,
    pub hints: f64,
    pub last_correct: SystemTime,
}

/// Question Enum
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash, Debug)]
pub enum QuestionVariant {
    Term(Term),
    List(List),
    Bullet(Bullet),
}

/// Flash Cards
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash, Debug)]
pub struct Term {
    term: String,
    definition: String,
    inverted: bool,
}

// Not a fan of this section having children, but I'll allow it for now
// Also, term? item? body? Pick one.
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash, Debug)]
pub struct List {
    order: u32,
    item: String,
    details: Vec<String>,
}

#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash, Debug)]
pub struct Bullet {
    body: String,
}

impl Term {
    pub fn new(term: String, definition: String) -> Question {
        Question::new(QuestionVariant::Term(Self {
            term,
            definition,
            inverted: false,
        }))
    }

    pub fn _flip(&mut self) {
        self.inverted = !self.inverted;
    }
}

impl List {
    // FIXME: Slice of strings here? &[] not Vec?
    pub fn new(order: u32, item: String, details: Vec<String>) -> Question {
        Question::new(QuestionVariant::List(Self {
            order,
            item,
            details,
        }))
    }
}

impl Bullet {
    pub fn new(body: String) -> Question {
        Question::new(QuestionVariant::Bullet(Self { body }))
    }
}

impl Question {
    pub fn new(data: QuestionVariant) -> Question {
        Question {
            id: Uuid::new_v4(),
            data,
            comp_level: Strictness::Trimmed,
            mastery: 0,
            correct: 0,
            seen: 0,
            hints: 0.0,
            last_correct: SystemTime::now(),
        }
    }

    pub fn ask(&self) -> &str {
        match &self.data {
            QuestionVariant::Term(t) => {
                if t.inverted {
                    &t.definition
                } else {
                    &t.term
                }
            }
            _ => todo!(),
        }
    }

    pub fn peek(&self) -> &str {
        match &self.data {
            QuestionVariant::Term(t) => {
                if t.inverted {
                    &t.term
                } else {
                    &t.definition
                }
            }
            _ => todo!(),
        }
    }

    pub fn answer(&mut self, ans: &str, hints: f64) -> (bool, &str) {
        let right_ans = self.peek().to_owned();
        let correct = logic::check_answer(ans, &right_ans, &self.comp_level);
        self.seen += 1;
        if correct {
            self.correct += 1;
            self.hints += hints;
            if hints == 0.0 {
                self.increment_mastery();
                self.last_correct = SystemTime::now();
            }
        } else {
            self.decrement_mastery();
        }
        (correct, self.peek())
    }

    pub fn override_correct(&mut self) {
        self.correct += 1;
        if self.correct > self.seen {
            self.correct = self.seen;
        } else {
            // Give back the mastery lost from the wrong answer
            self.increment_mastery();
            // Give the new mastery for being correct
            self.increment_mastery();
        }
    }

    pub fn _get_variant(&mut self) -> &mut QuestionVariant {
        &mut self.data
    }

    // FIXME: Make the range of min and max mastery configurable
    fn increment_mastery(&mut self) {
        if self.mastery < 10 {
            self.mastery += 1;
        }
    }

    fn decrement_mastery(&mut self) {
        if self.mastery > 0 {
            self.mastery -= 1;
        }
    }
}

impl PartialEq for Question {
    fn eq(&self, other: &Self) -> bool {
        self.data == other.data
    }
}

impl Eq for Question {}

impl Hash for Question {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.data.hash(state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_term(t: &str, d: &str) -> Question {
        Term::new(t.to_string(), d.to_string())
    }

    #[test]
    fn variant_equality() {
        let a = make_term("Bonjour", "Hello");
        let b = make_term("Bonjour", "Hello");
        assert_eq!(a, b);
        assert_ne!(a.id, b.id);
    }

    #[test]
    fn mastery_lower_bound() {
        let mut a = make_term("", "right");
        for _ in 1..100 {
            a.answer("wrong", 0.0);
        }
        assert_eq!(a.mastery, 0);
    }

    #[test]
    fn mastery_upper_bound() {
        let mut a = make_term("", "right");
        for _ in 1..100 {
            a.answer("right", 0.0);
        }
        assert_eq!(a.mastery, 10);
    }

    #[test]
    fn mastery_up_and_down() {
        let mut a = make_term("", "right");
        a.answer("right", 0.0);
        a.answer("right", 0.0);
        a.answer("right", 0.0);
        a.answer("wrong", 0.0);
        a.answer("right", 0.0);
        a.answer("wrong", 0.0);
        assert_eq!(a.mastery, 2);
    }

    #[test]
    fn last_correct_time() {
        let mut a = make_term("", "right");
        let t1 = a.last_correct.clone();
        a.answer("wrong", 0.0);
        a.answer("wrong", 0.0);
        let t2 = a.last_correct.clone();
        a.answer("right", 0.0);
        let t3 = a.last_correct.clone();
        assert_eq!(t1, t2);
        assert!(t3 > t2);
    }

    #[test]
    fn override_correct_works() {
        let mut a = make_term("", "right");
        a.answer("right", 0.0);
        a.answer("wrong", 0.0);
        assert!(a.correct < a.seen);
        assert_eq!(a.mastery, 0);
        a.override_correct();
        assert_eq!(a.correct, a.seen);
        assert_eq!(a.mastery, 2);
    }

    #[test]
    fn override_correct_cant_be_cheated() {
        let mut a = make_term("", "right");
        a.answer("right", 0.0);
        assert_eq!(a.correct, a.seen);
        assert_eq!(a.mastery, 1);
        a.override_correct();
        assert_eq!(a.correct, a.seen);
        assert_eq!(a.mastery, 1);
    }

    #[test]
    fn ask_term() {
        let a = make_term("question", "right");
        assert_eq!(a.ask(), "question");
    }

    #[test]
    fn peek_term() {
        let a = make_term("question", "right");
        assert_eq!(a.peek(), "right");
        assert_eq!(a.seen, 0);
    }
}
