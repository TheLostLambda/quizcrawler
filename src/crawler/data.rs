use crate::crawler::util::*;
use serde_derive::Deserialize;

/// Flash Cards
#[derive(Debug)]
pub struct Flash {
    term: String,
    definition: String,
}

impl Flash {
    pub fn new(t: &str, d: &str) -> Flash {
        Flash {
            term: t.to_owned(),
            definition: d.to_owned(),
        }
    }
}

/// Generic Result for using `?`
type BoxResult<T> = Result<T,Box<dyn std::error::Error>>;

/// This struct holds all of the configuration data that is parsed from the TOML
#[derive(Debug, Deserialize)]
pub struct Config {
    pub title: String,
    pub exts: Vec<String>,
    pub flash: FlashConfig,
}

impl Config {
    pub fn from_file(file_name: &str) -> BoxResult<Config> {
        let toml_str = read_file_as_string(file_name)?;
        Ok(toml::from_str(&toml_str)?)
    }
}
        
/// This struct holds the regex components needed to extract flashcards
#[derive(Debug, Deserialize)]
pub struct FlashConfig {
    pub leader: String,
    pub term: String,
    pub separator: String,
    pub definition: String,
    pub terminator: String,
}
