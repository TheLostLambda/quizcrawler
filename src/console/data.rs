use crate::core::{
    data::{Section, Path},
    quiz::{Dispatcher, MultipleChoice, QuizProgress, QuizRef, DSettings},
};
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use std::collections::HashMap;

// Trim back things that don't need to be public

#[derive(Default)]
pub struct QCSettings {}

pub struct Quizcrawler {
    pub tree: Section,
    pub settings: QCSettings,
    pub state_stack: Vec<State>,
}

#[derive(Default)]
pub struct TreeState {
    pub path: Path,
    pub selection_history: HashMap<Path, usize>,
}

impl TreeState {
    pub fn get_selected_mut(&mut self) -> &mut usize {
        self.selection_history.entry(self.path.clone()).or_insert(0)
    }

    pub fn get_selected(&self) -> usize {
        *self.selection_history.get(&self.path).unwrap_or(&0)
    }
}

#[derive(Clone)]
pub struct QuestionState {
    pub quiz: QuizRef,
    pub progress: QuizProgress,
}

// FIXME: This could use some more thought
pub enum State {
    TreeView(TreeState),
    Dispatch(Dispatcher),
    AskQuestion(QuestionState),
    AnswerQuestion(QuestionState, (bool, String)), // FIXME: Ew
}

impl Quizcrawler {
    pub fn new(settings: QCSettings, tree: Section) -> Self {
        Self {
            tree,
            settings,
            state_stack: vec![State::TreeView(TreeState::default())],
        }
    }

    pub fn handle_key(&mut self, key: KeyEvent) {
        match self.state_stack.last_mut() {
            Some(State::TreeView(state)) => {
                // FIXME: Maybe move this to a function?
                let node = self.tree.child_at_path(&state.path).unwrap(); // FIXME: Unwrap
                let child_names: Vec<_> = node.children.iter().map(|x| &x.name).collect();
                let limit = child_names.len() - 1;
                let current = state.get_selected();
                // This gives me an entry, a mutable reference which is updated in the match
                let selector = state.get_selected_mut();
                match key.code {
                    KeyCode::Char('q') => {
                        self.state_stack.pop();
                    }
                    KeyCode::Up if current > 0 => *selector -= 1,
                    KeyCode::Down if current < limit => *selector += 1,
                    KeyCode::Right if node.children[current].is_parent() => {
                        state.path.push(child_names[current].to_owned())
                    }
                    // FIXME: This feels a tad muddled
                    KeyCode::Char(' ') => {
                        let mut path = state.path.clone();
                        path.push(child_names[current].to_owned());
                        // FIXME: unwrap is likely a bad idea here
                        let section = self
                            .tree
                            .child_at_path(&path)
                            .unwrap();
                        let settings = DSettings {
                            recursive: key.modifiers.contains(KeyModifiers::CONTROL),
                            ..Default::default()
                        };
                        let mut dispatcher = Dispatcher::new(settings, section);
                        dispatcher.register_quiz(MultipleChoice::default());
                        self.state_stack.push(State::Dispatch(dispatcher))
                    }
                    KeyCode::Left => {
                        state.path.pop();
                    }
                    _ => {}
                }
            }
            Some(State::AskQuestion(state)) => {
                if !state.quiz.borrow().get_choices().is_empty() {
                    match key.code {
                        KeyCode::Char('q') => {
                            // FIXME: This double-pop is hacky and is needed to get past the dispatcher
                            // Write a rewind function to pop the stack back until some condition is met,
                            // like reaching something isn't Dispatch. Maybe the states could have a
                            // transient flag and all of those are popped off.
                            self.state_stack.pop();
                            self.state_stack.pop();
                        }
                        KeyCode::Char('h') => state.quiz.borrow_mut().get_hint(),
                        KeyCode::Char(c) => {
                            let result = state.quiz.borrow_mut().answer(&c.to_string());
                            if let Some(result) = result {
                                let state = state.clone();
                                self.state_stack.pop();
                                self.state_stack.push(State::AnswerQuestion(state, result))
                            }
                        }
                        _ => {}
                    }
                }
            }
            Some(State::AnswerQuestion(state, (correct, _))) => match key.code {
                KeyCode::Char('q') => {
                    // FIXME: Hacky double-pop
                    self.state_stack.pop();
                    self.state_stack.pop();
                }
                KeyCode::Char('o') if !*correct => {
                    state.quiz.borrow_mut().i_was_right();
                    *correct = true;
                }
                KeyCode::Char(' ') => {
                    self.state_stack.pop();
                }
                _ => {}
            },
            _ => {}
        }
    }

    pub fn tick(&mut self) -> bool {
        match self.state_stack.last_mut() {
            Some(State::Dispatch(dispatcher)) => {
                if let Some(quiz) = dispatcher.next() {
                    let progress = dispatcher.progress();
                    let state = QuestionState { quiz, progress };
                    self.state_stack.push(State::AskQuestion(state));
                } else {
                    self.state_stack.pop();
                }
                true
            }
            Some(_) => true,
            None => false,
        }
    }
}
