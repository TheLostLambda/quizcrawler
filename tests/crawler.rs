use insta::assert_ron_snapshot;
use quizcrawler::crawler::data::Crawler;
use std::fs;

const CONF_FILE: &str = "confs/borg.toml";

fn crawler() -> Crawler {
    let conf_str = fs::read_to_string(CONF_FILE).unwrap();
    Crawler::new(&conf_str).unwrap_or_else(|err| {
        panic!(
            "Failed to parse a config from string. The error was: {}\n The config string was: {}",
            err, conf_str
        );
    })
}

#[test]
fn test_parse_terms() {
    let data_str = r#"
* Terms
  - der / die Lehrer(in) :: teacher
  - wissen / weiÃŸ / hat gewusst :: to know
  - in der Zwischenzeit :: [in the] meantime"#;

    let cards = crawler().parse_sections(data_str);
    assert_ron_snapshot!(cards, {".**.atime" => "[atime]"});
}

#[test]
fn test_parse_lists() {
    let data_str = r#"
* Terms
  1) Use numbers to indicate some sort of process or ranking.
     - This is an annotation about this step in the process.
  2) This, for example, happens after the first point.
  3) Or perhaps this is the third most expensive solution."#;

    let cards = crawler().parse_sections(data_str);
    assert_ron_snapshot!(cards, {".**.atime" => "[atime]"});
}

#[test]
fn test_parse_bullets() {
    let data_str = r#"
* Terms
  - Here is some short, relevant fact regarding this subtopic.
  - And another one! Only use these when there is no better option.
  - Definitions and processes have their own structures."#;

    let cards = crawler().parse_sections(data_str);
    assert_ron_snapshot!(cards, {".**.atime" => "[atime]"});
}

#[test]
fn test_parse_multiline_terms() {
    let data_str = r#"
* Terms
  - Quizcrawler :: Quizcrawler is an application that, when fed a file of
    class-notes, crawls the structure and generates interactive quizzes that can be
    used as review. It leverages spaced repetition and forced / active recall to
    enhance learning. The gamification of studying should further increase
    engagement and recall.
"#;

    let cards = crawler().parse_sections(data_str);
    assert_ron_snapshot!(cards, {".**.atime" => "[atime]"});
}
