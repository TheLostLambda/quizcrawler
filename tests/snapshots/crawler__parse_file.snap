---
source: tests/crawler.rs
expression: section
---
Section(
  name: "borg.org",
  questions: [],
  children: [
    Section(
      name: "Professors",
      questions: [
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "Prof. Joe Generic (j.generic@sheffield.ac.uk)",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "Dr. Bob Assistant (b.assistant@sheffield.ac.uk)",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
      ],
      children: [],
    ),
    Section(
      name: "Resources",
      questions: [
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "Link to a textbook",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "A neat webpage",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "A relevant paper",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "Etc.",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
      ],
      children: [],
    ),
    Section(
      name: "Assignments",
      questions: [],
      children: [
        Section(
          name: "TODO [#A] This is an important and time-sensitive assignment!",
          questions: [],
          children: [],
        ),
        Section(
          name: "TODO [#B] This is less urgent but needs to done in a timely manner.",
          questions: [],
          children: [],
        ),
        Section(
          name: "TODO [#C] This is something that I\'d like to do but can wait.",
          questions: [],
          children: [],
        ),
      ],
    ),
    Section(
      name: "Notes",
      questions: [],
      children: [
        Section(
          name: "Theme I",
          questions: [],
          children: [
            Section(
              name: "Topic 1",
              questions: [
                Question(
                  id: "[id]",
                  data: Bullet(Bullet(
                    body: "This is information that is relevant to the topic as a whole",
                  )),
                  comp_level: Trimmed,
                  mastery: 0,
                  correct: 0,
                  seen: 0,
                  hints: 0,
                  last_correct: "[last_correct]",
                ),
                Question(
                  id: "[id]",
                  data: Bullet(Bullet(
                    body: "This section should be small. Try to use subtopics where possible.",
                  )),
                  comp_level: Trimmed,
                  mastery: 0,
                  correct: 0,
                  seen: 0,
                  hints: 0,
                  last_correct: "[last_correct]",
                ),
              ],
              children: [
                Section(
                  name: "Subtopic A",
                  questions: [
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Here is some short, relevant fact regarding this subtopic.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "And another one! Only use these when there is no better option.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Definitions and processes have their own structures.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                  ],
                  children: [],
                ),
                Section(
                  name: "Subtopic B",
                  questions: [
                    Question(
                      id: "[id]",
                      data: List(List(
                        order: 1,
                        item: "Use numbers to indicate some sort of process or ranking.",
                        details: [
                          "This is an annotation about this step in the process.",
                        ],
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: List(List(
                        order: 2,
                        item: "This, for example, happens after the first point.",
                        details: [],
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: List(List(
                        order: 3,
                        item: "Or perhaps this is the third most expensive solution.",
                        details: [],
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                  ],
                  children: [],
                ),
                Section(
                  name: "Subtopic C",
                  questions: [
                    Question(
                      id: "[id]",
                      data: Term(Term(
                        term: "Term",
                        definition: "Definition (This is for vocabulary)",
                        inverted: false,
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Term(Term(
                        term: "Term",
                        definition: "Another definition (sometimes these are better suited as subtopics)",
                        inverted: false,
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Term(Term(
                        term: "Term",
                        definition: "Def. (if so, don\'t use this format â€” just create a new heading)",
                        inverted: false,
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                  ],
                  children: [],
                ),
              ],
            ),
            Section(
              name: "Topic 2",
              questions: [],
              children: [
                Section(
                  name: "Subtopic A",
                  questions: [
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Some points regarding emphasis",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "/italic words/ are vocabulary words and are part of the topical vernacular.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "*Bold words* are buzzwords or important concepts. Key ideas are in bold.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "_Underlined words_ are proper nouns â€” like the names of organizations.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "~Code segments~ may be useful for courses to do with programming.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                  ],
                  children: [],
                ),
              ],
            ),
          ],
        ),
        Section(
          name: "Theme II",
          questions: [],
          children: [
            Section(
              name: "Topic 1",
              questions: [
                Question(
                  id: "[id]",
                  data: Bullet(Bullet(
                    body: "Avoid putting text at any level higher than Topic (3).",
                  )),
                  comp_level: Trimmed,
                  mastery: 0,
                  correct: 0,
                  seen: 0,
                  hints: 0,
                  last_correct: "[last_correct]",
                ),
              ],
              children: [
                Section(
                  name: "Subtopic A",
                  questions: [
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Here are a couple of general notes",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "You can, and sometimes should, have more than 1 layer of subtopics",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Avoid any bare text (nothing outside of ordered, unordered, or term lists)",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Images are allowed to be bare (without a preceding bullet)",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Never indent a list. If it needs more structure, split into subtopics.",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "Equations are subtopics with a bullet for LaTeX + variables : definitions",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                  ],
                  children: [],
                ),
                Section(
                  name: "Table B",
                  questions: [
                    Question(
                      id: "[id]",
                      data: Bullet(Bullet(
                        body: "FIXME: Eventually put a table here!",
                      )),
                      comp_level: Trimmed,
                      mastery: 0,
                      correct: 0,
                      seen: 0,
                      hints: 0,
                      last_correct: "[last_correct]",
                    ),
                  ],
                  children: [],
                ),
              ],
            ),
          ],
        ),
      ],
    ),
    Section(
      name: "Scratch",
      questions: [
        Question(
          id: "[id]",
          data: List(List(
            order: 1,
            item: "This",
            details: [],
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: List(List(
            order: 2,
            item: "Then",
            details: [],
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: List(List(
            order: 3,
            item: "That",
            details: [
              "[[file:.orgimg/Scratch/screenshot_2019-04-08_14-05-47.png]] You don\'t even need lists to structure things here. This space is just meant to be a buffer for content that eventually makes its way into the notes section. As such, headings aren\'t anything special here:",
            ],
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "This is a free space for dropping in-class notes",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
        Question(
          id: "[id]",
          data: Bullet(Bullet(
            body: "There is no real structure here, so previous rules don\'t apply",
          )),
          comp_level: Trimmed,
          mastery: 0,
          correct: 0,
          seen: 0,
          hints: 0,
          last_correct: "[last_correct]",
        ),
      ],
      children: [
        Section(
          name: "Here \'cause I want to be",
          questions: [],
          children: [],
        ),
      ],
    ),
  ],
)
